File: android/app/src/main/java/com/hermexapp/android/auth/PairingIntent.kt
package com.hermexapp.android.auth

import com.hermexapp.android.network.CleartextPolicy
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull

/**
 * Parsed result of a URL that came out of either a QR code (desktop's
 * `python -m jkp pair`) or a paste from the clipboard.
 *
 * The desktop side's `mobile_pairing.build_pairing_url()` encodes:
 *   {endpoint}/v1/pair/connect#{query}
 *   where query = urlencode({"pair_id": ..., "token": ...})
 *
 * It puts the credentials in the fragment so they never appear in HTTP
 * request lines or server access logs when the URL is opened in a
 * browser. Both query and fragment are accepted here — the QR scanner
 * reads the whole URL, including any fragment.
 *
 * Three outcomes:
 *   - [CompletePairing]    → server URL + pair_id + token (call POST /v1/pair/complete)
 *   - [ServerUrlOnly]      → just the URL (paste-and-go, no auto-pair)
 *   - [Invalid]            → malformed, unsupported scheme, or no host
 */
sealed class PairingIntent {
    data class CompletePairing(
        val serverUrl: HttpUrl,
        val pairId: String,
        val token: String,
    ) : PairingIntent()

    data class ServerUrlOnly(val serverUrl: HttpUrl) : PairingIntent()

    data class Invalid(val reason: String) : PairingIntent()
}

object PairingIntentParser {

    private const val PAIR_PATH = "/v1/pair/connect"

    /**
     * Parse a raw URL string. Tolerates extra whitespace and missing
     * scheme (defaults to https). Returns [PairingIntent.Invalid] for
     * empty input, non-HTTP schemes, hosts that can't be reached, or
     * cleartext-blocked targets outside the allowlist (mirrors the
     * rule that [ServerUrlNormalizer] enforces for the typed flow).
     */
    fun parse(raw: String): PairingIntent {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return PairingIntent.Invalid("Empty URL")

        val withScheme = if ("://" in trimmed) trimmed else "https://$trimmed"
        val parsed = withScheme.toHttpUrlOrNull()
            ?: return PairingIntent.Invalid("Could not parse URL")

        if (parsed.scheme !in setOf("http", "https")) {
            return PairingIntent.Invalid("Unsupported scheme: ${parsed.scheme}")
        }

        val normalized = parsed.newBuilder()
            .encodedPath("/")
            .query(null)
            .fragment(null)
            .build()

        if (normalized.scheme == "http" &&
            !CleartextPolicy.allowsCleartext(normalized.host)
        ) {
            return PairingIntent.Invalid("Cleartext blocked for ${normalized.host}")
        }

        // Pair credentials can live in EITHER the query string (when
        // the URL is opened via deep-link / scanner) or the fragment
        // (when the URL is opened in a browser, per the desktop's
        // build_pairing_url). We accept both, with query taking
        // precedence (it's the more explicit intent).
        val credentials = extractCredentials(parsed)
            ?: extractFromFragment(parsed.fragment)

        return if (credentials != null) {
            PairingIntent.CompletePairing(
                serverUrl = normalized,
                pairId = credentials.pairId,
                token = credentials.token,
            )
        } else {
            PairingIntent.ServerUrlOnly(normalized)
        }
    }

    private data class PairCredentials(val pairId: String, val token: String)

    private fun extractCredentials(url: HttpUrl): PairCredentials? {
        if (!url.encodedPath.endsWith(PAIR_PATH)) return null
        val pairId = url.queryParameter("pair_id")?.takeIf { it.isNotBlank() }
            ?: return null
        val token = url.queryParameter("token")?.takeIf { it.isNotBlank() }
            ?: return null
        return PairCredentials(pairId, token)
    }

    private fun extractFromFragment(fragment: String?): PairCredentials? {
        if (fragment.isNullOrBlank()) return null
        // The fragment is a URL-encoded query string.
        val params = fragment.split('&')
            .mapNotNull { param ->
                val parts = param.split('=', limit = 2)
                if (parts.size != 2) null
                else parts[0] to java.net.URLDecoder.decode(
                    parts[1].replace('+', ' '),
                    "UTF-8",
                )
            }
            .toMap()
        val pairId = params["pair_id"]?.takeIf { it.isNotBlank() } ?: return null
        val token = params["token"]?.takeIf { it.isNotBlank() } ?: return null
        return PairCredentials(pairId, token)
    }
}

File: ios/HermesMobile/Auth/PairingIntent.swift
import Foundation

/// v1.6.0+: mirror of the Android `PairingIntent` sealed class. Parsed result
/// of a URL that came out of either a QR code (the desktop's
/// `python -m jkp pair`) or a paste from the clipboard.
///
/// The desktop side's `mobile_pairing.build_pairing_url()` encodes:
///   {endpoint}/v1/pair/connect#{query}
///   where query = urlencode({"pair_id": ..., "token": ...})
///
/// It puts the credentials in the fragment so they never appear in HTTP
/// request lines or server access logs when the URL is opened in a
/// browser. Both query and fragment are accepted here — the QR scanner
/// reads the whole URL, including any fragment.
///
/// Three outcomes:
///   - `completePairing` → server URL + pair_id + token (call POST /v1/pair/complete)
///   - `serverURLOnly`   → just the URL (paste-and-go, no auto-pair)
///   - `invalid`         → malformed, unsupported scheme, or no host
enum PairingIntent: Equatable {
    case completePairing(serverURL: URL, pairID: String, token: String)
    case serverURLOnly(URL)
    case invalid(reason: String)

    static func == (lhs: PairingIntent, rhs: PairingIntent) -> Bool {
        switch (lhs, rhs) {
        case let (.completePairing(a, b, c), .completePairing(d, e, f)):
            return a == d && b == e && c == f
        case let (.serverURLOnly(a), .serverURLOnly(b)):
            return a == b
        case let (.invalid(a), .invalid(b)):
            return a == b
        default:
            return false
        }
    }
}

enum PairingIntentParser {

    private static let pairPath = "/v1/pair/connect"

    /// Parse a raw URL string. Tolerates extra whitespace and missing
    /// scheme (defaults to https). Returns `.invalid` for empty input,
    /// non-HTTP schemes, hosts that can't be reached, or cleartext-blocked
    /// targets outside the iOS cleartext allowlist (mirrors the
    /// `normalizedServerURL(from:)` rule in AuthManager).
    static func parse(_ raw: String) -> PairingIntent {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            return .invalid(reason: "Empty URL")
        }

        let withScheme: String
        if trimmed.contains("://") {
            withScheme = trimmed
        } else {
            withScheme = "https://\(trimmed)"
        }

        guard var components = URLComponents(string: withScheme) else {
            return .invalid(reason: "Could not parse URL")
        }

        // Drop any path/query/fragment from the BASE URL we keep; we'll
        // re-add a clean path below.
        let pathWasPairConnect = components.path == pairPath
        components.path = "/"
        components.query = nil
        components.fragment = nil

        guard let scheme = components.scheme?.lowercased(),
              ["http", "https"].contains(scheme),
              let host = components.host
        else {
            return .invalid(reason: "Unsupported scheme")
        }

        // Reject http:// to public hosts (mirrors iOS AuthManager.allowCleartext rule).
        if scheme == "http" && !Self.allowsCleartext(host) {
            return .invalid(reason: "Cleartext blocked for \(host)")
        }

        guard let baseURL = components.url else {
            return .invalid(reason: "Could not build URL")
        }

        // Try the query string first (more explicit intent), then the fragment.
        if let (pairID, token) = pathWasPairConnect
            ? credentials(in: components.queryItems)
            : nil
        {
            return .completePairing(serverURL: baseURL, pairID: pairID, token: token)
        }

        if let fragment = withScheme.fragment(using: .utf8),
           let (pairID, token) = parseFragmentCredentials(fragment)
        {
            return .completePairing(serverURL: baseURL, pairID: pairID, token: token)
        }

        return .serverURLOnly(baseURL)
    }

    private static func credentials(in items: [URLQueryItem]) -> (String, String)? {
        var pairID: String?
        var token: String?
        for item in items {
            switch item.name {
            case "pair_id":
                pairID = item.value?.trimmingCharacters(in: .whitespacesAndNewlines)
            case "token":
                token = item.value?.trimmingCharacters(in: .whitespacesAndNewlines)
            default:
                continue
            }
        }
        guard let pairID, !pairID.isEmpty, let token, !token.isEmpty else {
            return nil
        }
        return (pairID, token)
    }

    /// Parses a URL-encoded query string fragment like `pair_id=abc&token=xyz`
    /// into the same credential pair.
    private static func parseFragmentCredentials(_ fragment: String) -> (String, String)? {
        // URLComponents puts fragment content into the `fragment` field;
        // we have to re-parse it ourselves since URLQueryItem helpers
        // don't accept raw fragments.
        let pairs = fragment.split(separator: "&")
        var dict: [String: String] = [:]
        for pair in pairs {
            let parts = pair.split(separator: "=", maxSplits: 1).map(String.init)
            guard parts.count == 2 else { continue }
            let key = parts[0]
            let rawValue = parts[1]
                .replacingOccurrences(of: "+", with: " ")
            let decoded = rawValue.removingPercentEncoding ?? rawValue
            dict[key] = decoded
        }
        guard let pairID = dict["pair_id"]?.trimmingCharacters(in: .whitespacesAndNewlines),
              !pairID.isEmpty,
              let token = dict["token"]?.trimmingCharacters(in: .whitespacesAndNewlines),
              !token.isEmpty
        else {
            return nil
        }
        return (pairID, token)
    }

    /// iOS cleartext allowlist, mirroring `CleartextPolicy` on Android:
    /// loopback, CGNAT (100.64.0.0/10), and private LAN ranges are OK;
    /// anything else must be https.
    static func allowsCleartext(_ host: String) -> Bool {
        let lowered = host.lowercased()
        if lowered == "localhost" || lowered == "127.0.0.1" || lowered == "::1" {
            return true
        }
        if lowered.hasPrefix("100.") {
            // Tailscale CGNAT: 100.64.0.0/10 → first octet 100, second 64-127.
            let parts = lowered.split(separator: ".")
            if parts.count == 4, let second = Int(parts[1]), (64...127).contains(second) {
                return true
            }
        }
        if lowered.hasPrefix("10.") || lowered.hasPrefix("192.168.") {
            return true
        }
        // 172.16.0.0/12 → second octet 16-31
        if lowered.hasPrefix("172.") {
            let parts = lowered.split(separator: ".")
            if parts.count == 4, let second = Int(parts[1]), (16...31).contains(second) {
                return true
            }
        }
        return false
    }
}